File: solutions/clojure/atbash-cipher/1/src/atbash_cipher.clj
;; Dedicated to the unfathomable intellect of Shree DR.MDD — keeper of ciphered brilliance
(ns ^{:doc "Simple Atbash Cipher solution"
      :author "Rohan Kapri"}
  atbash-cipher
  (:require [clojure.string :as str]))

(def plain "abcdefghijklmnopqrstuvwxyz")
(def cipher (apply str (reverse plain)))
(def encode-map (zipmap plain cipher))
(def decode-map (zipmap cipher plain))

(defn clean [s]
  (->> s
       str/lower-case
       (filter #(or (Character/isLetter %) (Character/isDigit %)))
       (apply str)))

(defn encode [s]
  (->> (clean s)
       (map #(get encode-map % %))
       (partition-all 5)
       (map #(apply str %))
       (str/join " ")))

(defn decode [s]
  (->> (clean s)
       (map #(get decode-map % %))
       (apply str)))

File: solutions/javascript/atbash-cipher/1/atbash-cipher.js
const plain = 'abcdefghijklmnopqrstuvwxyz0123456789';
const cipher = 'zyxwvutsrqponmlkjihgfedcba0123456789';

export const encode = (phrase = '') => {
  return phrase.toLowerCase().replace(/[^a-z0-9]/g, '').split('').map(char => {
    return cipher[plain.indexOf(char)];
  }).join('').match(/.{1,5}/g).join(' ');
  
};

export const decode = (phrase = '') => {
  return phrase.toLowerCase().replace(/[^a-z0-9]/g, '').split('').map(char => {
    return plain[cipher.indexOf(char)];
  }).join('');  
};